sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/LockRenewalOperation.java

```diff
@@ -0,0 +1,212 @@
+        return status.get();
+    }
+
+    /**
+     * Gets the exception if an error occurred whilst renewing the message or session lock.
+     *
+     * @return the exception if an error occurred whilst renewing the message or session lock, otherwise {@code null}.
+     */
+    public Throwable getThrowable() {
+        return throwable.get();
+    }
+
+    /**
+     * Cancels the lock renewal operation.
+     */
+    @Override
+    public void close() {
+        if (isDisposed.getAndSet(true)) {
+            return;
+        }
+
+        if (status.compareAndSet(LockRenewalStatus.RUNNING, LockRenewalStatus.CANCELLED)) {
+            logger.verbose("token[{}] Cancelled operation.", lockToken);
+        }
+
+        cancellationProcessor.onComplete();
+        subscription.dispose();
+    }
+
+    /**
+     * Gets the lock renewal operation. if the {@code maxLockRenewalDuration} is {@link Duration#isZero()}, then the
+     * lock is never renewed.
+     *
+     * @param initialLockedUntil When the initial call is locked until.
+     * @param maxLockRenewalDuration Duration to renew lock for.
+     * @return The subscription for the operation.
+     */
+    private Disposable getRenewLockOperation(Instant initialLockedUntil, Duration maxLockRenewalDuration) {
+        if (maxLockRenewalDuration.isZero()) {
+            status.set(LockRenewalStatus.COMPLETE);
+            return Disposables.single();
+        }
+
+        final Instant now = Instant.now();
+        Duration initialInterval = Duration.between(now, initialLockedUntil);
+
+        if (initialInterval.isNegative()) {
+            logger.info("Duration was negative. now[{}] lockedUntil[{}]", now, initialLockedUntil);
+            initialInterval = Duration.ZERO;
+        } else {
+            // Adjust the interval, so we can buffer time for the time it'll take to refresh.
+            final Duration adjusted = MessageUtils.adjustServerTimeout(initialInterval);
+            if (adjusted.isNegative()) {
+                logger.info("Adjusted duration is negative. Adjusted: {}ms", initialInterval.toMillis());
+            } else {
+                initialInterval = adjusted;
+            }
+        }
+
+        final EmitterProcessor<Duration> emitterProcessor = EmitterProcessor.create();
+        final FluxSink<Duration> sink = emitterProcessor.sink();
+
+        sink.next(initialInterval);
+
+        final Flux<Object> cancellationSignals = Flux.first(cancellationProcessor, Mono.delay(maxLockRenewalDuration));
+
+        return Flux.switchOnNext(emitterProcessor.map(Flux::interval))
+            .takeUntilOther(cancellationSignals)
+            .flatMap(delay -> {
+                final String id = lockToken;
+                logger.info("token[{}]. now[{}]. Starting lock renewal.", id, Instant.now());
+                if (CoreUtils.isNullOrEmpty(id)) {
+                    return Mono.error(new IllegalStateException("Cannot renew session lock without session id."));
```

Comment:
is this messaging correct ?, the lockToken can be for Message Lock or session id.